sycl/source/detail/persistent_cache.cpp

```diff
@@ -0,0 +1,259 @@
+//==---------- persistent_cache.cpp - On-disk cache for program -*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <detail/config.hpp>
+#include <detail/persistent_cache.hpp>
+#include <detail/plugin.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+/* Stores build program in persisten cache
+ */
+void PersistentCache::putPIProgramToDisc(const detail::plugin &Plugin,
+                                         const device &Device,
+                                         const RTDeviceBinaryImage &Img,
+                                         const SerializedObj &SpecConsts,
+                                         const std::string &BuildOptionsString,
+                                         const RT::PiProgram &Program) {
+
+  if (!isPersistentCacheEnabled())
+    return;
+
+  // Only SPIRV images are cached
+  if (Img.getFormat() != PI_DEVICE_BINARY_TYPE_SPIRV &&
+      (Img.getFormat() == PI_DEVICE_BINARY_TYPE_NONE &&
+       pi::getBinaryImageFormat(Img.getRawData().BinaryStart, Img.getSize()) !=
+           PI_DEVICE_BINARY_TYPE_SPIRV))
+    return;
+
+  std::string DirName =
+      getCacheItemDirName(Device, Img, SpecConsts, BuildOptionsString);
+
+  size_t i = 0;
+  std::string FileName;
+  do {
```

Comment:
Another general comment is about synchronization - I don't see any. Is it a TODO, or the persistent code cache is thread-safe?

Reply:
Persistent cache is behind in-memory cache (the read/write calls happens in build step which is already synchronized).

Reply:
Does this reliably work when multiple SYCL apps are running in parallel?

Reply:
> Does this reliably work when multiple SYCL apps are running in parallel?

It is in TODO list (see PR description). Next patch should have inter-process synchronization mechanism for FS access.

Reply:
I see. Please add a comment about per-process safety due to lock-step with in-memory cache and a TODO to add inter-proc sync.